### ISSUE TITLE
Connecting to external linstor storage with piraeus-operator on the okd cluster failed

### ISSUE THREAD

**Issue by lts9527** [piraeusdatastore/piraeus-operator#450]
Environment

  oc version

```
Client Version: 4.12.0-0.okd-2023-03-18-084815
Kustomize Version: v4.5.7
Server Version: 4.12.0-0.okd-2023-03-18-084815
Kubernetes Version: v1.25.0-2786+eab9cc98fe4c00-dirty
```

  

   helm list

```
NAME    NAMESPACE       REVISION        UPDATED                                 STATUS          CHART           APP VERSION
piraeus default         2               2023-04-03 12:22:32.426185285 +0000 UTC deployed        piraeus-1.10.2  1.10.2     
```

  helm installation command
  
  ```
  helm install piraeus ./charts/piraeus \
      --values \
      ./charts/piraeus/values.yaml,./charts/piraeus/values-openshift.yaml \
      --set etcd.enabled=false \
      --set operator.controller.dbConnectionURL=k8s \
      --set operator.controller.enabled=false \
      --set csi.enableTopology=false \
      --set controllerEndpoint=http://[xxx]:3370
  ```

status

kubectl get pod

```
NAME                                         READY   STATUS                  RESTARTS        AGE
piraeus-csi-controller-8c69d6c74-x29wj       6/6     Running                 0               46m
piraeus-csi-node-9vtpz                       0/3     Init:0/1                0               46m
piraeus-csi-node-fsphb                       0/3     Init:0/1                0               46m
piraeus-ns-node-bhfpl                        0/2     Init:CrashLoopBackOff   13 (34s ago)    42m
piraeus-ns-node-wfx5d                        0/2     Init:CrashLoopBackOff   13 (17s ago)    42m
piraeus-operator-65bc8f99cb-49248            1/1     Running                 0               46m

```

Related Tips

```
time="2023-04-03T13:01:10Z" level=debug msg="reconcile error list" Name=piraeus-ns Namespace=default controller=LinstorSatelliteSet
time="2023-04-03T13:01:10Z" level=info msg="satellite Reconcile: reconcile loop end" Controller=linstorsatelliteset controller=LinstorSatelliteSet err="multiple errors: online satellite registered by operator without associated k8s node" requestName=piraeus-ns requestNamespace=default result="{false 0s}"
2023-04-03T13:01:10Z    ERROR   Reconciler error        {"controller": "LinstorSatelliteSet-controller", "object": {"name":"piraeus-ns","namespace":"default"}, "namespace": "default", "name": "piraeus-ns", "reconcileID": "98dd2f0e-972b-41c5-922a-9dd4c9a9fe5b", "error": "multiple errors: online satellite registered by operator without associated k8s node"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:329
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:274
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:235
time="2023-04-03T13:01:15Z" level=info msg="Reconciling LinstorCSIDriver" requestName=piraeus requestNamespace=default
time="2023-04-03T13:01:15Z" level=debug msg="reconcile spec with env" requestName=piraeus requestNamespace=default
time="2023-04-03T13:01:15Z" level=debug msg="reconcile spec with resources" requestName=piraeus requestNamespace=default
time="2023-04-03T13:01:15Z" level=debug msg="performing upgrades and fill defaults in resource" Name=piraeus Namespace=default Op=reconcileResource
time="2023-04-03T13:01:15Z" level=debug msg="performing upgrade/fill: #1 -> Set default image names for CSI" Name=piraeus Namespace=default Op=reconcileResource
time="2023-04-03T13:01:15Z" level=debug msg="finished upgrade/fill: #1 -> Set default image names for CSI: changed=false" Name=piraeus Namespace=default Op=reconcileResource


time="2023-04-03T13:05:15Z" level=info msg="Reconciling LinstorCSIDriver" requestName=piraeus requestNamespace=default
time="2023-04-03T13:05:15Z" level=debug msg="reconcile spec with env" requestName=piraeus requestNamespace=default
time="2023-04-03T13:05:15Z" level=debug msg="reconcile spec with resources" requestName=piraeus requestNamespace=default
time="2023-04-03T13:05:15Z" level=debug msg="performing upgrades and fill defaults in resource" Name=piraeus Namespace=default Op=reconcileResource
time="2023-04-03T13:05:15Z" level=debug msg="performing upgrade/fill: #1 -> Set default image names for CSI" Name=piraeus Namespace=default Op=reconcileResource
time="2023-04-03T13:05:15Z" level=debug msg="finished upgrade/fill: #1 -> Set default image names for CSI: changed=false" Name=piraeus Namespace=default Op=reconcileResource
time="2023-04-03T13:05:15Z" level=debug msg="performing upgrade/fill: #2 -> Set default endpoint URL for client" Name=piraeus Namespace=default Op=reconcileResource
time="2023-04-03T13:05:15Z" level=debug msg="finished upgrade/fill: #2 -> Set default endpoint URL for client: changed=false" Name=piraeus Namespace=default Op=reconcileResource
time="2023-04-03T13:05:15Z" level=debug msg="performing upgrade/fill: #3 -> Set service account names to previous implicit values" Name=piraeus Namespace=default Op=reconcileResource
time="2023-04-03T13:05:15Z" level=debug msg="finished upgrade/fill: #3 -> Set service account names to previous implicit values: changed=false" Name=piraeus Namespace=default Op=reconcileResource
time="2023-04-03T13:05:15Z" level=debug msg="performing upgrade/fill: #4 -> Set kubelet path to default" Name=piraeus Namespace=default Op=reconcileResource
time="2023-04-03T13:05:15Z" level=debug msg="finished upgrade/fill: #4 -> Set kubelet path to: changed=false" Name=piraeus Namespace=default Op=reconcileResource
time="2023-04-03T13:05:15Z" level=debug msg="finished all upgrades/fills" Name=piraeus Namespace=default Op=reconcileResource
time="2023-04-03T13:05:15Z" level=debug msg="creating csi node daemon set" Name=piraeus Namespace=default Op=reconcileNodes
time="2023-04-03T13:05:15Z" level=debug msg="reconciling csi node objects" Name=piraeus Namespace=default Op=reconcileNodes
time="2023-04-03T13:05:15Z" level=debug msg="creating linstor client" Name=piraeus Namespace=default Op=reconcileCSINodeObjects
time="2023-04-03T13:05:15Z" level=debug msg="curl -X 'GET' -H 'Accept: application/json' 'http://[controllerEndpoint]:3370/v1/controller/version'"
time="2023-04-03T13:05:15Z" level=debug msg="fetching linstor nodes" Name=piraeus Namespace=default Op=reconcileCSINodeObjects
time="2023-04-03T13:05:15Z" level=debug msg="curl -X 'GET' -H 'Accept: application/json' 'http://[controllerEndpoint]:3370/v1/nodes'"
time="2023-04-03T13:05:15Z" level=debug msg="searching matching linstor node" pod=piraeus-csi-node-fsphb
time="2023-04-03T13:05:15Z" level=debug msg="no linstor node found, skipping" pod=piraeus-csi-node-fsphb
time="2023-04-03T13:05:15Z" level=debug msg="searching matching linstor node" pod=piraeus-csi-node-9vtpz
time="2023-04-03T13:05:15Z" level=debug msg="no linstor node found, skipping" pod=piraeus-csi-node-9vtpz
time="2023-04-03T13:05:15Z" level=debug msg="creating csi controller deployment" Name=piraeus Namespace=default Op=reconcileControllerDeployment
time="2023-04-03T13:05:15Z" level=debug msg="creating csi driver resource" Name=piraeus Namespace=default Op=reconcileCSIDriver

```

**Comment by lts9527** [comment]
I think I found the reason
```
kubectl logs piraeus-ns-node-fv5jm
Defaulted container "linstor-satellite" out of: linstor-satellite, drbd-prometheus-exporter, kernel-module-injector (init)
```
I found something similar here. https://github.com/piraeusdatastore/piraeus-operator/issues/404

**Comment by WanzenBug** [comment]
See also https://github.com/piraeusdatastore/piraeus-operator/blob/master/doc/host-setup.md#injector-image-for-compiling-without-headers-on-host

There is also a new project for compiling kernel modules in Kubernetes: https://kmm.sigs.k8s.io/
You might be able to use that, and use `operator.satelliteSet.kernelModuleInjectionMethod=DepsOnly`. You can find the sources for DRBD here: https://pkg.linbit.com/

**Comment by lts9527** [comment]
Thank you very much for your reply to help me solve the problem

There have been many twists and turns in the process.

My kernel version
uname -r
```
6.1.11-200.fc37.x86_64
```
I need to replace the mirror version of the dockerfile 
```
quay.io/piraeusdatastore/drbd9-centos8:v9.2.2
```

否则使用drbd-9.1.12 会出现的错误
```
/tmp/pkg/drbd-9.1.12/drbd/drbd_dax_pmem.c: In function 'drbd_dax_open':
/tmp/pkg/drbd-9.1.12/drbd/drbd_dax_pmem.c:62:19: error: too few arguments to function 'fs_dax_get_by_bdev'
   62 |         dax_dev = fs_dax_get_by_bdev(bdev->md_bdev);
      |                   ^~~~~~~~~~~~~~~~~~
In file included from /tmp/pkg/drbd-9.1.12/drbd/drbd_dax_pmem.c:24:
./include/linux/dax.h:134:20: note: declared here
  134 | struct dax_device *fs_dax_get_by_bdev(struct block_device *bdev, u64 *start_off,
      |                    ^~~~~~~~~~~~~~~~~~
make[3]: *** [scripts/Makefile.build:250: /tmp/pkg/drbd-9.1.12/drbd/drbd_dax_pmem.o] Error 1
make[2]: *** [Makefile:2009: /tmp/pkg/drbd-9.1.12/drbd] Error 2
make[1]: *** [Makefile:134: kbuild] Error 2
make[1]: Leaving directory '/tmp/pkg/drbd-9.1.12/drbd'
make: *** [Makefile:126: module] Error 2
```
